Mathport/Syntax/Translate/Basic.lean

```diff
@@ -168,37 +206,109 @@ def renameModule (n : Name) : M Name := do Rename.renameModule (← read).pcfg n
 def renameField (n : Name) : M Name := return Rename.renameField? (← getEnv) n |>.getD n
 def renameOption (n : Name) : M Name := warn! "warning: unsupported option {n}" | pure n
 
-def mkIdentI (n : Name) (choices : Array Name := #[]) : M Syntax :=
-  return mkIdent (← renameIdent n choices)
-def mkIdentA (n : Name) : M Syntax := mkIdent <$> renameAttr n
-def mkIdentN (n : Name) : M Syntax := mkIdent <$> renameNamespace n
-def mkIdentF (n : Name) : M Syntax := mkIdent <$> renameField n
-def mkIdentO (n : Name) : M Syntax := mkIdent <$> renameOption n
+def mkIdentR (n : Name) : M Syntax := return (mkIdent n).setInfo (← MonadRef.mkInfoFromRefPos)
+
+def mkIdentI (n : Name) (choices : Array Name := #[]) : M Syntax := do mkIdentR (← renameIdent n choices)
+def mkIdentA (n : Name) : M Syntax := do mkIdentR (← renameAttr n)
+def mkIdentN (n : Name) : M Syntax := do mkIdentR (← renameNamespace n)
+def mkIdentF (n : Name) : M Syntax := do mkIdentR (← renameField n)
+def mkIdentO (n : Name) : M Syntax := do mkIdentR (← renameOption n)
 
 def Parser.ParserM.run' (p : ParserM α) (args : Array (Spanned VMCall)) : M α := do
   match p.run ⟨(← read).commands, args⟩ with
   | Except.ok a => pure a
   | Except.error e => throw! "unsupported: {e}"
 
+def mkCommentString (comment : Comment) : String :=
+  if comment.text.contains '\n' then s!"/-{comment.text}-/\n" else s!"--{comment.text}\n"
```

Comment:
What if you have a comment like `def foo /- not bar -/ := 1`? Using a line comment would cause other things to get commented out.

Reply:
There's a newline at the end of the string, so it would be:
```lean
def foo -- not bar
  := 1
```
The formatting wouldn't be better with `/- -/` since we also add a newline at the end.  The formatting options for comments are really limited.  The heuristic here seems to work okay on the files I've tested though.

Reply:
This is an actual output.  I guess we're not annotating the binder names with the correct span, so the comment is attached to the binder type instead.  But I'm going to wait for the results on mathlib before tweaking this.

```lean
unsafe axiom dsimplify_core
    {α :-- The user state type.
      Type}
    (a :-- Initial user data
      α)
    (pre :/- (pre a e) is invoked before visiting the children of subterm 'e',
             if it succeeds the result (new_a, new_e, flag) where
               - 'new_a' is the new value for the user data
               - 'new_e' is a new expression that must be definitionally equal to 'e',
               - 'flag'  if tt 'new_e' children should be visited, and 'post' invoked. -/
        α →
        expr → tactic (α × expr × Bool))
    (post :/- (post a e) is invoked after visiting the children of subterm 'e',
             The output is similar to (pre a e), but the 'flag' indicates whether
             the new expression should be revisited or not. -/
        α →
        expr → tactic (α × expr × Bool))
    (e : expr) (cfg : DsimpConfig := {  }) : tactic (α × expr)
```

Reply:
One actually ugly example from mathlib:
```lean
inductive prequotient
  |
  of :-- There's always `of`
    ∀ j : J x : F.obj j, prequotient
  |
  zero :-- Then one generator for each operation
    prequotient
```

The original is this:
```lean
inductive prequotient
-- There's always `of`
| of : Π (j : J) (x : F.obj j), prequotient
-- Then one generator for each operation
| zero : prequotient
```

Reply:
Another similar issue:
```lean
instance : LaxMonoidal.{u} (free R).obj where
  ε :=-- Send `R` to `punit →₀ R`
      ε
      R
  μ :=-- Send `(α →₀ R) ⊗ (β →₀ R)` to `α × β →₀ R`
      μ
      R
```

Reply:
Is the lack of space before the comment caused by this code or by the formatter? Maybe you should just throw a space in `addTrailingComment'`.

I'm not sure where the binder span issue is, it seems like the `unspanning` in `trBinders'` should do the trick. Perhaps the problem is that you are attaching the comment to the first generated syntax in left-to-right inside-out order, instead of outside-in. At least, I don't think it's a problem with the spans themselves but with the way the comments are getting attached.

Reply:
I'm currently fixing these issues, and I'm done with structure instances and declarations binders.  The `unspanning` didn't help much there because the binder nodes are constructed manually.

Reply:
> Is the lack of space before the comment caused by this code or by the formatter?

That's from the formatter, or rather from stuff like `":" ppSpace` I believe.  Adding a space only helps in these cases where the comment is already at the wrong position anyhow, the common case becomes more ugly.

Reply:
Another ugly effect comes from the parser for inductive constructors:
```lean
def ctor             := leading_parser "\n| " >> ppIndent (declModifiers true >> ident >> optional inferMod >> optDeclSig)
```
The `\n` in the first token means that the comment is printed *before* the newline and not before the pipe, as one would expect:
```lean
  | trans : ∀ x y z h : relation x y k : relation y z, relation x z-- There's always a `map` relation

  | map :
    ∀ j j' : J f : j ⟶ j' x : F.obj j,
      relation (of j' (F.map f x)) (of j x)-- Then one relation per operation, describing the interaction with `of`

  | zero : ∀ j, relation (of j 0) zero
```

Anyhow, I believe these are best fixed in core.  We are now putting the comments on the right tokens, it's just the formatter(s) that need tweaking.

Reply:
The issues in `instance ... where` is because the `mathport_rules` which creates that syntax doesn't preserve the field span (because it is pattern matched using `$[$k:ident := $v:term $[,]?]*`, so it's not trivial to get the field syntax here).